### DISCUSSION TITLE
Repeater interface hopes to add file fields

### DISCUSSION THREAD

**Discussion by cdwmhcc** [directus/directus#8834] [Archive]
![image](https://user-images.githubusercontent.com/8030134/137540589-a1ce8557-6f55-4140-ac32-e3ac1e269be0.png)

**Comment by rijkvanzanten** [comment]
The JSON repeater can't do relational values. Use a one-to-many or many-to-many instead 👍🏻

**Reply by Kematia** [reply]
How feasible would it be to have it work in repeaters as well? I reckon most of us would be fine losing the relational data for a much leaner json stored data.

e: seeing that in this case we only store the ID of the image

**Reply by rijkvanzanten** [reply]
It would mean we'd have to add some sort of field into the repeater configuration to fake the relationship information (instead of relying on the database), and then we'd have to rebuild every relational interface to work with this new fake relationship info, as it's not coming from the database anymore. The relational interfaces also currently allow for opening items in nested drawers, which is again done by following the relationships 🤔 

It's technically not impossible, but it basically means we'd have to reinvent all the relational stuff to work non-relationally, in which case you're honestly better off just using a o2m/m2m instead. The output JSON from the API would be identical, but you lose the technical overhead, and you gain all the other powerful features you get from relational values, like deep nesting, sorting, deep filtering etc etc etc 🤔 

TL;DR: Possible, but difficult. Not really worth it IMO

**Reply by benhaynes** [reply]
That's what extensions are for! Shouldn't be too hard for someone to make a non-relational file/image interface (that stores the asset URL), and then it can be used. That's just a "workaround" that won't be shipped in core.

**Reply by Kematia** [reply]
I'm thinking more: "just store the ID of the image, I don't care about relationships and I will eventually handle it on my own on the frontend" . What you're describing is having a json value/interface behave a 100% like a database relation, which I don't think is what most people want... I think some (including me) might want to trade off relationality for the ergonomics of a JSON object. 

The most common use case is having a "gallery/slider" interface in an article, which, at least in my opinion, doesn't warrant setting up a more "cumbersome" o2m relation (and what if I want to store data with that slide? Like a title or a link? I now have to setup a m2m with additional fields...) my mindset is: JSON for quick, cheap data, and relations for more important data structures. It's the developer choice what to setup when, and I don't think that directus should get into the way.

**Reply by benhaynes** [reply]
First, if you set up a M2M for those slideshow images, you can add "context" fields on the junction for Title, Link, etc. We also will be making a dedicated Gallery (multiple images) interface soon... though this will also be relational. The purpose of Directus is to make those relationships **not feel cumbersome**, not to remove relationships from a relational database. 😉 

Second, everything you're describing is possible... just via extensions. The goal of Directus **Core** is to cover ~80% of use-cases, no more than that to avoid bloating the codebase. The other 20% is handled by extensions. So again, you could create a special repeater to handle this... or just an image/file interface that is non-relational. Both of those will get you exactly what you're looking for. 💯 

However, if we add this into Core, I guarantee you that within 1 week someone will ask for that repeater full of images to start including nested image data in the response... which we can't do. As always: K.I.S.S. 💋

**Reply by rijkvanzanten** [reply]
>  I guarantee you that within 1 week someone will ask for that repeater full of images to start including nested image data in the response

100% this ☝🏻 

would also make it very difficult to document; "You can use `fields` to fetch nested relational data, _except_ for when you're using a JSON field, even though it looks the same, it's not the same"

**Reply by Kematia** [reply]
> Second, everything you're describing is possible... just via extensions.

I know! I just like arguing 😂

**Reply by alexbrowng** [reply]
I understand both parts and they seem to me very valid, but, perhaps there is a middle way. In a relational database, relationships are the most optimal way to relate information, generally, the reason for using JSON fields is not usually to establish relationships, it is rather to store data without defining a fixed data schema, in the case of the API the requests of a JSON field defined with the repeating interface and a relational field can be accessed practically the same and in the case of the App is really where you experience different, if you could create a relationship using the repeating interface that would half cover the two perspectives, for me the repeating interface is very good and I find the relationship interface a little more complex.

**Comment by benhaynes** [comment]
Merging this into: https://github.com/directus/directus/discussions/3001
